nexus/db-queries/src/db/datastore/switch_port.rs

```diff
@@ -829,45 +830,158 @@ impl DataStore {
-                    .map_err(|e| {
-                        public_error_from_diesel(e, ErrorHandler::Server)
-                    })?;
-            }
-            UpdatePrecondition::Null => {
-                diesel::update(switch_port_dsl::switch_port)
-                    .filter(switch_port::id.eq(switch_port_id))
-                    .filter(switch_port::port_settings_id.is_null())
-                    .set(switch_port::port_settings_id.eq(port_settings_id))
-                    .execute_async(&*conn)
-                    .await
-                    .map_err(|e| {
-                        public_error_from_diesel(e, ErrorHandler::Server)
-                    })?;
-            }
-            UpdatePrecondition::Value(current_id) => {
-                diesel::update(switch_port_dsl::switch_port)
-                    .filter(switch_port::id.eq(switch_port_id))
-                    .filter(switch_port::port_settings_id.eq(current_id))
-                    .set(switch_port::port_settings_id.eq(port_settings_id))
-                    .execute_async(&*conn)
-                    .await
-                    .map_err(|e| {
-                        public_error_from_diesel(e, ErrorHandler::Server)
-                    })?;
-            }
-        }
+        let err = OptionalError::new();
+        self.transaction_retry_wrapper("switch_port_set_settings_id")
+            .transaction(&conn, |conn| {
+                let err = err.clone();
+                async move {
+                    // TODO: remove once per-switch-multi-asn support is added
+                    // Bail if user attempts to assign multiple ASNs to a switch via switch port settings
+                    // This is a temporary measure until multi-asn-per-switch is supported.
+
+                    // what switch are we adding a configuration to?
+                    let switch = switch_port_dsl::switch_port
+                        .filter(switch_port_dsl::id.eq(switch_port_id))
+                        .select(switch_port_dsl::switch_location)
+                        .limit(1)
+                        .first_async::<String>(&conn)
+                        .await
+                        .map_err(|e: diesel::result::Error| {
+                            let msg = "failed to look up switch port by id";
+                            error!(opctx.log, "{msg}"; "error" => ?e);
+                            match e {
+                                diesel::result::Error::NotFound => {
+                                    err.bail(Error::not_found_by_id(
+                                        ResourceType::SwitchPort,
+                                        &switch_port_id,
+                                    ))
+                                }
+                                _ => err.bail(Error::internal_error(msg)),
+                            }
+                        })?;
+
+                    // if we're setting a port settings id (and therefore activating a configuration
```

Comment:
Good catch! I guess this is one of the consequences of being able to build up arbitrary configs, but only have some of them active - we have to validate that anything that is being activated is valid against the rest of the active set.